app/uk/gov/hmrc/hecstubs/controllers/AccountingPeriodController.scala

```diff
@@ -0,0 +1,111 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.hecstubs.controllers
+
+import cats.data.Validated.{Invalid, Valid}
+import cats.data.{Validated, ValidatedNel}
+import cats.syntax.apply._
+import com.google.inject.{Inject, Singleton}
+import play.api.libs.json.Json
+import play.api.mvc.{Action, AnyContent, ControllerComponents}
+import uk.gov.hmrc.hecstubs.models.accountOverviewDetails.InvalidCode._
+import uk.gov.hmrc.hecstubs.models.accountOverviewDetails.ReturnStatus.ReturnFound
+import uk.gov.hmrc.hecstubs.models.accountOverviewDetails.{Environment, ErrorResponse, ErrorResult}
+import uk.gov.hmrc.hecstubs.models.companyAccountingPeriod.{AccountingPeriod, CTUTR, CompanyAccountingPeriod}
+import uk.gov.hmrc.hecstubs.util.{Logging, ValidationUtils}
+import uk.gov.hmrc.play.bootstrap.backend.controller.BackendController
+
+import java.time.LocalDate
+import java.util.UUID
+
+@Singleton
+class AccountingPeriodController @Inject() (cc: ControllerComponents) extends BackendController(cc) with Logging {
+
+  def accountingPeriod(ctutr: String, startDate: String, endDate: String): Action[AnyContent] = Action { request =>
+    val correlationId = request.headers.get("CorrelationId").getOrElse(UUID.randomUUID().toString)
+    val environment   = request.headers.get("Environment").flatMap(Environment.getEnvironmentFromString)
+    environment match {
+      case Some(_) => validatedDataResponse(ctutr, startDate, endDate, correlationId)
+      case None    =>
+        logger.info(
+          s"Call failed to fetch  Company Accounting Period details as Environment in header is invalid"
+        )
+        BadRequest
+    }
+  }
+
+  private def validatedDataResponse(ctutr: String, startDate: String, endDate: String, correlationId: String) =
+    validateInputData(ctutr, startDate, endDate, correlationId) match {
+      case Invalid(errorList) =>
+        val errorResponse = ErrorResponse(errorList.toList)
+        logger.info(
+          s"Responding to call for Company Accounting Period failed with error Responses : ${errorResponse.toString()}"
+        )
+        BadRequest(Json.toJson(errorResponse))
+
+      case Valid(data) =>
+        val responseJson = Json.toJson(data)
+        logger.info(
+          s"Responding to call for Company Accounting Period  for UTR $ctutr , startDate: $startDate and  endDate: $endDate with JSON: ${responseJson.toString()}"
+        )
+        Ok(responseJson)
+    }
+
+  @SuppressWarnings(Array("org.wartremover.warts.Any"))
+  private def validateInputData(
+    ctutr: String,
+    startDate: String,
+    endDate: String,
+    correlationId: String
+  ): ValidatedNel[ErrorResult, CompanyAccountingPeriod] = {
+
+    val message    = "Submission has not passed validation."
```

Comment:
tiny bit of optimisation here, but this `val` and the regexes below it could come out of this function so that they're not being created every time the function is called

Reply:
sure 👍 . I was thinking about it and then forgot 🤦‍♀️

Reply:
updated